packages/compile-solidity/src/compilerSupplier/index.ts

```diff
@@ -2,114 +2,70 @@ import path from "path";
+      docker: Docker.Specification;
+      local: Local.Specification;
+      native: Native.Specification;
+      "version-range": VersionRange.Specification;
+    };
+  };
+}
 
-    let strategy;
-    const useDocker = this.docker;
-    const useNative = userSpecification === "native";
-    const useSpecifiedLocal =
-      userSpecification && this.fileExists(userSpecification);
-    const isValidVersionRange = semver.validRange(userSpecification);
+export const createCompilerSupplier = Common.CompilerSupplier.forDefinition<
+  CompilerSupplier.Specification
+>({
+  determineStrategy(options) {
+    const {
+      solcConfig: { docker = false, version }
+    } = options;
 
-    if (useDocker) {
-      strategy = new Docker(this.strategyOptions);
-    } else if (useNative) {
-      strategy = new Native();
-    } else if (useSpecifiedLocal) {
-      strategy = new Local();
-    } else if (isValidVersionRange) {
-      strategy = new VersionRange(this.strategyOptions);
+    if (docker) {
+      return "docker";
     }
 
-    if (strategy) {
-      const solc = await strategy.load(userSpecification);
-      return { solc };
-    } else {
-      throw this.badInputError(userSpecification);
+    if (version === "native") {
+      return "native";
     }
-  }
-
-  async list() {
-    const userSpecification = this.version;
-
-    let strategy;
-    const useDocker = this.docker;
-    const useNative = userSpecification === "native";
-    const useSpecifiedLocal =
-      userSpecification && this.fileExists(userSpecification);
-    const isValidVersionRange = semver.validRange(userSpecification);
 
-    if (useDocker) {
-      strategy = new Docker(this.strategyOptions);
-    } else if (useNative) {
-      strategy = new Native();
-    } else if (useSpecifiedLocal) {
-      strategy = new Local();
-    } else if (isValidVersionRange) {
-      strategy = new VersionRange(this.strategyOptions);
+    if (version && fileExists(version)) {
+      return "local";
     }
 
-    if (!strategy) {
-      throw this.badInputError(userSpecification);
+    if (!version || semver.validRange(version)) {
+      return "version-range";
     }
 
-    if (!strategy.list) {
-      throw new Error(
-        `Cannot list versions for strategy ${strategy.constructor.name}`
-      );
-    }
-
-    return await strategy.list();
-  }
-
-  static getDefaultVersion() {
-    return defaultSolcVersion;
-  }
-
-  badInputError(userSpecification) {
     const message =
-      `Could not find a compiler version matching ${userSpecification}. ` +
+      `Could not find a compiler version matching ${version}. ` +
       `compilers.solc.version option must be a string specifying:\n` +
       `   - a path to a locally installed solcjs\n` +
       `   - a solc version or range (ex: '0.4.22' or '^0.5.0')\n` +
       `   - a docker image name (ex: 'stable')\n` +
       `   - 'native' to use natively installed solc\n`;
-    return new Error(message);
+    throw new Error(message);
```

Comment:
As @haltman-at [mentions](https://github.com/trufflesuite/truffle/pull/4222#discussion_r687276676), this likely warrants its own error class to conform with the rest of this module.